### PR TITLE
[Console] Move AddConsoleCommandPass from FrameworkBundle to Console.

### PR DESCRIPTION
**Pull request by bcremer**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | yes |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | #19440 |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

**Comment by mcfedr** [comment]
Maybe worth merging (or reviewing) https://github.com/symfony/symfony/pull/19305 first so that it doesnt conflict.

File: src/Symfony/Bundle/FrameworkBundle/composer.json [hunk 1]
```diff
@@ -22,6 +22,7 @@
         "symfony/class-loader": "~3.2",
         "symfony/dependency-injection": "~3.2",
         "symfony/config": "~2.8|~3.0",
+        "symfony/console": "~3.2",
```
Comment (dunglas) [inline src/Symfony/Bundle/FrameworkBundle/composer.json]: This can be removed. Because of the class_exists call, if the installed component is an older version or isn't installed it will still work. Maybe should you add a another test to not load the deprecated pass if the component is not installed at all (to prevent throwing an illegitimate deprecation warning).

File: src/Symfony/Component/Console/DependencyInjection/AddConsoleCommandPass.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Console\DependencyInjection;
+
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+/**
+ * AddConsoleCommandPass.
```
Comment (dunglas) [inline src/Symfony/Component/Console/DependencyInjection/AddConsoleCommandPass.php]: This description isn't explicit at all (probably just something autogenerated by PHPStorm). It should be improved.

File: src/Symfony/Component/Console/Tests/DependencyInjection/AddConsoleCommandPassTest.php [hunk 3]
```diff
@@ -9,9 +9,9 @@
  * file that was distributed with this source code.
```
Comment (dunglas) [inline src/Symfony/Component/Console/Tests/DependencyInjection/AddConsoleCommandPassTest.php]: The deprecated pass must still be tested (the test will be removed at the same time than the class).

Comment (bcremer) [inline src/Symfony/Component/Console/Tests/DependencyInjection/AddConsoleCommandPassTest.php]: Reverted the test.

File: src/Symfony/Component/Console/DependencyInjection/AddConsoleCommandPass.php (hunk 2 above)
Comment (bcremer) [inline src/Symfony/Component/Console/DependencyInjection/AddConsoleCommandPass.php]: Added a more descriptive class description.

File: src/Symfony/Bundle/FrameworkBundle/composer.json (hunk 1 above)
Comment (bcremer) [inline src/Symfony/Bundle/FrameworkBundle/composer.json]: Removed the dependency.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/AddConsoleCommandPass.php [hunk 4]
```diff
@@ -15,12 +15,19 @@
...
  *
  * @author Grégoire Pineau <lyrixx@lyrixx.info>
+ *
+ * @deprecated since version 3.2, to be removed in 4.0. Use Symfony\Component\Console\DependencyInjection\AddConsoleCommandPass instead.
  */
 class AddConsoleCommandPass implements CompilerPassInterface
 {
+    public function __construct()
+    {
+        @trigger_error(sprintf('%s is deprecated as of 3.2 and will be removed in 4.0. Use Symfony\Component\Console\DependencyInjection\AddConsoleCommandPass instead.', __CLASS__), E_USER_DEPRECATED);
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/AddConsoleCommandPass.php]: this should be jus below the namespace declaration, see other places in the code base

File: src/Symfony/Bundle/FrameworkBundle/FrameworkBundle.php [hunk 5]
```diff
@@ -13,7 +13,7 @@
 
 use Symfony\Bundle\FrameworkBundle\DependencyInjection\Compiler\AddConstraintValidatorsPass;
 use Symfony\Bundle\FrameworkBundle\DependencyInjection\Compiler\AddValidatorInitializersPass;
-use Symfony\Bundle\FrameworkBundle\DependencyInjection\Compiler\AddConsoleCommandPass;
+use Symfony\Bundle\FrameworkBundle\DependencyInjection\Compiler\AddConsoleCommandPass as DeprecatedAddConsoleCommandPass;
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/FrameworkBundle.php]: we usually use the "Legacy" prefix here, instead of "Deprecated"

Comment (bcremer) [inline src/Symfony/Bundle/FrameworkBundle/FrameworkBundle.php]: Changed to Legacy.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/AddConsoleCommandPass.php (hunk 4 above)
Comment (bcremer) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/AddConsoleCommandPass.php]: Good point, changed.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/AddConsoleCommandPass.php [hunk 6]
```diff
@@ -11,13 +11,17 @@
 
 namespace Symfony\Bundle\FrameworkBundle\DependencyInjection\Compiler;
 
+@trigger_error(sprintf('%s is deprecated as of 3.2 and will be removed in 4.0. Use Symfony\Component\Console\DependencyInjection\AddConsoleCommandPass instead.', __CLASS__), E_USER_DEPRECATED);
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/AddConsoleCommandPass.php]: AddConsoleCommandPass::class instead of `__CLASS__` (which is undefined)

File: src/Symfony/Component/Console/composer.json [hunk 7]
```diff
@@ -20,12 +20,15 @@
...
     "require-dev": {
+        "symfony/http-kernel": "~2.8|~3.0",
         "symfony/event-dispatcher": "~2.8|~3.0",
+        "symfony/dependency-injection": "~2.8|~3.0",
         "symfony/filesystem": "~2.8|~3.0",
         "symfony/process": "~2.8|~3.0",
         "psr/log": "~1.0"
     },
     "suggest": {
+        "symfony/dependency-injection": "",
```
Comment (stof) [inline src/Symfony/Component/Console/composer.json]: I would remove this suggestion. The DI component does not allow to add more features to the Console component.

The compiler pass is useful only for people already using the component. So the suggestion is just useless noise.

Comment (bcremer) [inline src/Symfony/Component/Console/composer.json]: Makes sense; changed.

**Comment by fabpot** [comment]
Tests do not pass because of the deprecation.

**Comment by bcremer** [comment]
Tests are now all fixed.

File: UPGRADE-3.2.md [hunk 8]
```diff
@@ -6,6 +6,7 @@ FrameworkBundle
 
  * The `Controller::getUser()` method has been deprecated and will be removed in
    Symfony 4.0; typehint the security user object in the action instead.
+ * The `Symfony\Bundle\FrameworkBundle\DependencyInjection\Compiler\AddConsoleCommandPass` has been deprecated. Use `Symfony\Component\Console\DependencyInjection\AddConsoleCommandPass` instead.
```
Comment (xabbuh) [inline UPGRADE-3.2.md]: Please also add this to the `UPGRADE-4.0.md` (but reword it to "removed" there).

File: src/Symfony/Component/Console/Tests/DependencyInjection/AddConsoleCommandPassTest.php [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        $container->addCompilerPass(new AddConsoleCommandPass());
+
+        $definition = new Definition('SplObjectStorage');
+        $definition->addTag('console.command');
+        $container->setDefinition('my-command', $definition);
+
+        $container->compile();
+    }
+
+    public function testHttpKernelRegisterCommandsIngoreCommandAsAService()
```
Comment (xabbuh) [inline src/Symfony/Component/Console/Tests/DependencyInjection/AddConsoleCommandPassTest.php]: This whole test case looks a bit suspicious to me. We don't have anything HttpKernel specific here, do we?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/AddConsoleCommandPass.php [hunk 10]
```diff
@@ -11,14 +11,18 @@
...
 
 /**
- * AddConsoleCommandPass.
+ * Registers console commands.
  *
  * @author Grégoire Pineau <lyrixx@lyrixx.info>
+ *
+ * @deprecated since version 3.2, to be removed in 4.0. Use Symfony\Component\Console\DependencyInjection\AddConsoleCommandPass instead.
  */
 class AddConsoleCommandPass implements CompilerPassInterface
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/AddConsoleCommandPass.php]: Wouldn't it be better to let this class extend the new compiler pass instead of having to maintain the same code twice?

File: src/Symfony/Component/Console/Tests/DependencyInjection/AddConsoleCommandPassTest.php (hunk 9 above)
Comment (bcremer) [inline src/Symfony/Component/Console/Tests/DependencyInjection/AddConsoleCommandPassTest.php]: The test was copied 1 to 1 from the original location 
`\Symfony\Bundle\FrameworkBundle\Tests\DependencyInjection\Compiler\AddConsoleCommandPassTest::testHttpKernelRegisterCommandsIngoreCommandAsAService`.

Any suggestions how to name the testcase instead?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/AddConsoleCommandPass.php (hunk 10 above)
Comment (bcremer) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/AddConsoleCommandPass.php]: This would introduce a new dependency `symfony/console` to `symfony/framework-bundle` that is not yet defined in composer.json. 

Should the new dependency be added to composer.json or can should this fail during runtime?

Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/AddConsoleCommandPass.php]: We could add a conflict rule for versions of the Console component before 3.2.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/AddConsoleCommandPass.php [hunk 11]
```diff
@@ -11,14 +11,18 @@
 
 namespace Symfony\Bundle\FrameworkBundle\DependencyInjection\Compiler;
 
+@trigger_error(sprintf('%s is deprecated as of 3.2 and will be removed in 4.0. Use Symfony\Component\Console\DependencyInjection\AddConsoleCommandPass instead.', AddConsoleCommandPass::class), E_USER_DEPRECATED);
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/AddConsoleCommandPass.php]: [...] deprecated since version 3.2 [...]

**Comment by xabbuh** [comment]
👍 (failure is unrelated)

Status: Reviewed

File: src/Symfony/Component/Console/Tests/DependencyInjection/AddConsoleCommandPassTest.php (hunk 9 above)
Comment (xabbuh) [inline src/Symfony/Component/Console/Tests/DependencyInjection/AddConsoleCommandPassTest.php]: see #20273

**Comment by ro0NL** [comment]
Should we make some hardcoded id's / strings configurable thru required constructor args, as done in #20250? See discussion: https://github.com/symfony/symfony/pull/20250#discussion_r84143705

**Comment by bcremer** [comment]
Is there still anything I can help to get this merged into 3.3?

**Comment by xabbuh** [comment]
@bcremer You rebase this once again to resolve conflicts, but otherwise this looks ready to me.

ping @symfony/deciders

File: UPGRADE-3.2.md [hunk 12]
```diff
@@ -27,6 +27,7 @@ FrameworkBundle
  * The `Resources/public/images/*` files have been removed.
  * The `Resources/public/css/*.css` files have been removed (they are now inlined
    in TwigBundle).
+ * The `Symfony\Bundle\FrameworkBundle\DependencyInjection\Compiler\AddConsoleCommandPass` has been deprecated. Use `Symfony\Component\Console\DependencyInjection\AddConsoleCommandPass` instead.
```
Comment (xabbuh) [inline UPGRADE-3.2.md]: Ah sorry, now that Symfony 3.2 is released we would need to move this to the `UPGRADE-3.2md` file and update deprecation messages accordingly.

**Comment by bcremer** [comment]
@xabbuh Rebased on `master` and targeted to Symfony 3.3.

**Comment by fabpot** [comment]
Thank you @bcremer.
